sdks/java/io/kinesis/src/main/java/org/apache/beam/sdk/io/kinesis/ShardReadersPool.java

```diff
@@ -159,4 +212,91 @@ ShardRecordsIterator createShardIterator(SimplifiedKinesisClient kinesis,
+   *   <li>0 successors - when shard was merged with another shard and this one is considered
+   *   adjacent by merge operation</li>
+   *   <li>1 successor - when shard was merged with another shard and this one is considered a
+   *   parent by merge operation</li>
+   *   <li>2 successors - when shard was split into two shards</li>
+   * </ul>
+   * </p>
+   * <p>
+   * Once shard successors are established, the transition to reading new shards can begin.
+   * The {@link ShardReadersPool#shardIteratorsMap}, {@link ShardReadersPool#recordsQueue}
+   * and {@link ShardReadersPool#numberOfRecordsInAQueueByShard} need to be updated atomically to
+   * remain in sync. Therefore the operation is synchronized on this {@link ShardReadersPool}
+   * instance. During this operation, the immutable {@link ShardReadersPool#shardIteratorsMap}
+   * is replaced with a new one holding references to {@link ShardRecordsIterator} instances for
+   * open shards only. The {@link ShardReadersPool#recordsQueue} is replaced with a new instance of
+   * capacity adjusted to the new number of open shards and records remaining in previous queue are
+   * drained to the new one. Also, the counter for already closed shard is removed from
+   * {@link ShardReadersPool#numberOfRecordsInAQueueByShard} map.
+   * </p>
+   * <p>
+   * Finally when atomic update is finished, new threads are spawned for reading the successive
+   * shards. The thread that handled reading from already closed shard can finally complete.
+   * </p>
+   */
+  private void readFromSuccessiveShards(final ShardRecordsIterator closedShardIterator)
+      throws TransientKinesisException {
+    List<ShardRecordsIterator> successiveShardRecordIterators = closedShardIterator
+        .findSuccessiveShardRecordIterators();
+
+    // Handle one closedShardIterator at a time to keep shardIteratorsMap,
+    // recordsQueue and numberOfRecordsInAQueueByShard in sync.
+    synchronized (this) {
+      ImmutableMap.Builder<String, ShardRecordsIterator> shardsMap = ImmutableMap.builder();
+      Iterable<ShardRecordsIterator> allShards = Iterables
+          .concat(shardIteratorsMap.get().values(), successiveShardRecordIterators);
+      for (ShardRecordsIterator iterator : allShards) {
+        if (!closedShardIterator.getShardId().equals(iterator.getShardId())) {
+          shardsMap.put(iterator.getShardId(), iterator);
+        }
+      }
+      shardIteratorsMap.set(shardsMap.build());
+      numberOfRecordsInAQueueByShard.remove(closedShardIterator.getShardId());
+
+      BlockingQueue<KinesisRecord> previousRecordsQueue = recordsQueue.get();
+      int capacity = queueCapacityPerShard * shardIteratorsMap.get().size();
+      if (capacity > 0) {
+        BlockingQueue<KinesisRecord> newRecordsQueue = new ArrayBlockingQueue<>(capacity);
+        recordsQueue.set(newRecordsQueue);
+        do {
+          try {
+            KinesisRecord record = previousRecordsQueue.poll(500, TimeUnit.MILLISECONDS);
```

Comment:
Why do we need to poll here - don't we need to just transfer existing contents, which can be done using .drainTo()? If new elements can be added to this queue concurrently, then this loop is incorrect anyway, because it may exit at line 294 just before a thread adds a new element, and the element will be lost.

Reply:
drainTo() is tricky because it will throw exception when the new queue is full. The safer option is the combination of poll & put with blocking. There's a chance (rather little) of adding new elements to previousRecordsQueue while the reference to the new queue is set in recordsQueue atomic reference. That's why I've added polling with timeout just for this case. 
If the draining is stopped by InterruptedException then there's no need to do anything with remaining records. Threads in this pool can be interrupted only when the executionService is shutting down and this happens only after calling the stop() method.
I've added more comments in the code to make it more clear.

Reply:
Polling with timeout does not solve the issue - it can still happen that a record will be lost in limbo (added to the old queue at a moment when we've already switched to the new queue).

Imagine the following:
- recordsQueue is empty (all records so far have been already returned).
- A readLoop() thread (call it "thread A") calls recordsQueue.get() and gets a pointer to the old records queue.
- Another readLoop() thread ("thread B") sees a closed shard, runs the current code, and comes to line 286. Suppose for these 500ms there's still nothing in the queue e.g. because the OS didn't schedule thread A yet.
- At line 294 there still is nothing in the queue so we exit the loop.
- thread A wakes up and adds a record to the old records queue.

The record is lost: this is a data loss bug.